### PR TITLE
Jennyf/claims challenge/#712

### PR DESCRIPTION
**Pull request by jennyf19**
#712 
Claims challenge could be returned from middle tier service if it's doing downstream obo. Adal needs to update to 1) add new overloads to take in the claims challenge 2) skip cache lookup. 3) send passed in claims challenge to authorize endpoint

### CODE REVIEW

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 1]
```diff
@@ -79,9 +81,24 @@ public AcquireTokenInteractiveHandler(RequestData requestData, Uri redirectUri,
...
             this.DisplayableId = userId.DisplayableId;
             this.UserIdentifierType = userId.Type;
-            this.LoadFromCache = (requestData.TokenCache != null && parameters != null && PlatformPlugin.PlatformInformation.GetCacheLoadPolicy(parameters));
             this.SupportADFS = true;
 
+            claims = ProcessClaims(extraQueryParameters, claims);
+            if (!String.IsNullOrEmpty(claims))
+            {
+                PlatformPlugin.Logger.Verbose(CallState,
+                string.Format(CultureInfo.InvariantCulture,
```
Comment (aadidgit) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: MINOR Remove this formatting call and simply use the input string. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS3457)

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 2]
```diff
@@ -44,11 +44,13 @@ internal class AcquireTokenInteractiveHandler : AcquireTokenHandlerBase
 
         private readonly string extraQueryParameters;
 
+        private readonly string claims;
```
Comment (aadidgit) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: MAJOR Remove the unused private field 'claims'. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS1144)

File: src/ADAL.PCL/AuthenticationContext.cs [hunk 3]
```diff
@@ -556,6 +569,23 @@ public Guid CorrelationId
             return await handler.RunAsync().ConfigureAwait(false);
         }
 
+        private async Task<AuthenticationResult> AcquireTokenWithClaimsCommonAsync(string resource, ClientKey clientKey, Uri redirectUri, IPlatformParameters parameters,
```
Comment (aadidgit) [inline src/ADAL.PCL/AuthenticationContext.cs]: MAJOR Method has 8 parameters, which is greater than the 7 authorized. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS107)

File: src/ADAL.PCL/Http/AdalHttpClient.cs [hunk 4]
```diff
@@ -105,12 +111,29 @@ public AdalHttpClient(string uri, CallState callState)
                         serviceEx = new AdalServiceException(tokenResponse.Error, tokenResponse.ErrorDescription,
                             errorCodes, ex);
 
+                        if ((int)ex.WebResponse.StatusCode == 400 && tokenResponse.Error == interactionRequired)
```
Comment (aadidgit) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: CRITICAL Refactor this code to not nest more than 3 control flow statements. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS134)

File: src/ADAL.PCL.Android/BrokerProxy.cs [hunk 5]
```diff
@@ -430,6 +430,13 @@ private Bundle GetBrokerOptions(AuthenticationRequest request)
                 request.Version);
             brokerOptions.PutString(BrokerConstants.AccountExtraQueryParam,
                 request.ExtraQueryParamsAuthentication);
+
+            if(request.Claims != null)
+            {
+                brokerOptions.PutString(BrokerConstants.SkipCache, Boolean.TrueString.ToLower());
```
Comment (aadidgit) [inline src/ADAL.PCL.Android/BrokerProxy.cs]: MINOR Define the locale to be used in this string operation. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS1449)

**Comment by aadidgit** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/ADAL.PCL.Android/BrokerHelper.cs [hunk 6]
```diff
@@ -111,7 +111,7 @@ public void AcquireToken(IDictionary<string, string> brokerPayload)
 
                 // Don't send background request, if prompt flag is always or
                 // refresh_session
-                if (!string.IsNullOrEmpty(request.BrokerAccountName) || !string.IsNullOrEmpty(request.UserId))
+                if (string.IsNullOrEmpty(request.Claims) && !string.IsNullOrEmpty(request.BrokerAccountName) || !string.IsNullOrEmpty(request.UserId))
```
Comment (mjcheetham) [inline src/ADAL.PCL.Android/BrokerHelper.cs]: Did you mean this to be..
```cs
(A && !B) || !C
```
..or..
```cs
A && (!B || !C)

Currently it's the former.
```

**Comment by duncanpMS** [comment]
@jennyf19 the failing _sonarqube_ check won't block you from merging the changes to _dev_ (only the _AppVeyor_ checks will block merges).  The _sonarqube_ check is running static analysis rules and reporting issues only in code that you have added or changed i.e. new issues you have introduced. It's there as an automated check to help during reviews: it saves time for reviewers by automatically commenting on common issues and catches things they might miss.
Just treat them as suggestions and decide whether they need to fixed or not (some are more important than others).

File: src/ADAL.PCL/Http/AdalHttpClient.cs [hunk 7]
```diff
@@ -105,12 +111,29 @@ public AdalHttpClient(string uri, CallState callState)
...
+                        {
+                            // Extracts the error and claims data from exception
+                            string temp = ex.InnerException.InnerException.Message;
```
Comment (mjcheetham) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: Please use a better variable name than `temp`, e.g., `errorJson`
😄

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 8]
```diff
@@ -79,9 +81,24 @@ public AcquireTokenInteractiveHandler(RequestData requestData, Uri redirectUri,
...
+                    "Claims present. Skip cache lookup."));
+            }
+            this.LoadFromCache = (requestData.TokenCache != null && parameters != null && PlatformPlugin.PlatformInformation.GetCacheLoadPolicy(parameters) && String.IsNullOrEmpty(claims));
+
+            // Push claims into extraQueryParameters.
+            if (!String.IsNullOrEmpty(this.extraQueryParameters))
+            {
+                this.extraQueryParameters += "&";
+            }
+            this.extraQueryParameters += "claims=" + claims;
```
Comment (mjcheetham) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: The content of `claims` comes from the user (developer). Do we need to URL encode these before appending them to the query params string?

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 9]
```diff
@@ -247,5 +265,51 @@ protected override bool BrokerInvocationRequired()
 
             return false;
         }
+
+        private string ProcessClaims(string extraQueryParameters, string claims)
```
Comment (mjcheetham) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: What does this method do? Maybe a brief comment as to what 'processing' is done here.
Also are there no Framework types (or existing library helpers) which might help here with parsing/merging query strings?

File: samples/AdalAndroidTestApp/MainActivity.cs [hunk 10]
```diff
@@ -133,6 +136,31 @@ protected override void OnActivityResult(int requestCode, Result resultCode, Int
...
+            this.accessTokenTextView.Text = string.Empty;
+            TokenBroker tokenBroker = new TokenBroker();
+            tokenBroker.Sts = sts;
+            EditText email = FindViewById<EditText>(Resource.Id.email);
+            tokenBroker.Sts.ValidUserName = email.Text;
+            string value = null;
+            try
+            {
+                string policy = "{\"access_token\":{\"polids\":{\"essential\":true,\"values\":[\"5ce770ea-8690-4747-aa73-c5b3cd509cd4\"]}}}";
+                value = await tokenBroker.GetTokenInteractiveAsync(new PlatformParameters(this, true), null, policy);
```
Comment (HSAR) [inline samples/AdalAndroidTestApp/MainActivity.cs]: Are you planning to add an E2E test for this flow?

File: src/ADAL.PCL/Http/AdalHttpClient.cs (hunk 4 above)
Comment (duncanpMS) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: There's no need to cast to _int_ - you could just compare to the defined enum value _HttpStatusCode.BadRequest_.

File: src/ADAL.PCL.Android/BrokerProxy.cs (hunk 5 above)
Comment (jennyf19) [inline src/ADAL.PCL.Android/BrokerProxy.cs]: done.

File: src/ADAL.PCL.Android/BrokerHelper.cs (hunk 6 above)
Comment (jennyf19) [inline src/ADAL.PCL.Android/BrokerHelper.cs]: Should be the latter. Updated.

File: src/ADAL.PCL/Http/AdalHttpClient.cs (hunk 7 above)
Comment (jennyf19) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: done.

File: samples/AdalAndroidTestApp/MainActivity.cs (hunk 10 above)
Comment (jennyf19) [inline samples/AdalAndroidTestApp/MainActivity.cs]: I've tested E2E for Android with Broker, and OBO with a .NET web service. More E2E tests are in progress

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 11]
```diff
@@ -79,9 +81,24 @@ public AcquireTokenInteractiveHandler(RequestData requestData, Uri redirectUri,
...
             this.UserIdentifierType = userId.Type;
-            this.LoadFromCache = (requestData.TokenCache != null && parameters != null && PlatformPlugin.PlatformInformation.GetCacheLoadPolicy(parameters));
             this.SupportADFS = true;
 
+            claims = ProcessClaims(extraQueryParameters, claims);
+            if (!String.IsNullOrEmpty(claims))
+            {
+                PlatformPlugin.Logger.Verbose(CallState,
+                string.Format(CultureInfo.InvariantCulture,
+                    "Claims present. Skip cache lookup."));
```
Comment (jennyf19) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: Great feedback and explanation. Thank you.

File: src/ADAL.PCL/Http/AdalHttpClient.cs [hunk 12]
```diff
@@ -42,6 +45,9 @@ class AdalHttpClient
         private const string PKeyAuthName = "PKeyAuth";
         private const int DelayTimePeriodMilliSeconds = 1000;
 
+        // Error Constants
+        const String interactionRequired = "interaction_required";
```
Comment (jennyf19) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: done

File: src/ADAL.PCL/Http/AdalHttpClient.cs (hunk 4 above)
Comment (jennyf19) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: done.

File: src/ADAL.PCL/Utilities/InteractionRequiredExceptionDetails.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System.Runtime.Serialization;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory.Utilities
+{
+    [DataContract]
+    public class InteractionRequiredExceptionDetails
```
Comment (jennyf19) [inline src/ADAL.PCL/Utilities/InteractionRequiredExceptionDetails.cs]: done.

File: tests/Test.ADAL.NET.Unit/ClaimsChallengeExceptionTests.cs [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler()
+            {
+                Method = HttpMethod.Post,
+                ResponseMessage = new HttpResponseMessage(HttpStatusCode.BadRequest)
+                {
+                    Content = new StringContent(claims)
+                }
+            });
+
+            try
```
Comment (jennyf19) [inline tests/Test.ADAL.NET.Unit/ClaimsChallengeExceptionTests.cs]: Great suggestion...this makes things much clearer.

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs (hunk 9 above)
Comment (jennyf19) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: done.

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs (hunk 8 above)
Comment (jennyf19) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: I don't think this is necessary. We're not sending claims as eqp. I'm removing this part.

File: src/ADAL.PCL/Flows/AcquireTokenHandlerBase.cs [hunk 15]
```diff
@@ -120,6 +117,8 @@ protected AcquireTokenHandlerBase(RequestData requestData)
 
                 if (this.LoadFromCache)
                 {
+                    PlatformPlugin.Logger.Verbose(CallState, string.Format(CultureInfo.InvariantCulture,
```
Comment (aadidgit) [inline src/ADAL.PCL/Flows/AcquireTokenHandlerBase.cs]: MINOR Remove this formatting call and simply use the input string. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS3457)

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 16]
```diff
@@ -79,9 +81,23 @@ public AcquireTokenInteractiveHandler(RequestData requestData, Uri redirectUri,
...
             this.UserIdentifierType = userId.Type;
-            this.LoadFromCache = (requestData.TokenCache != null && parameters != null && PlatformPlugin.PlatformInformation.GetCacheLoadPolicy(parameters));
             this.SupportADFS = true;
 
+            claims = ProcessClaims(extraQueryParameters, claims);
+            if (!String.IsNullOrEmpty(claims))
+            {
+                this.LoadFromCache = false;
+
+                PlatformPlugin.Logger.Verbose(CallState, string.Format(CultureInfo.InvariantCulture,
```
Comment (aadidgit) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: MINOR Remove this formatting call and simply use the input string. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS3457)

File: src/ADAL.PCL/Http/AdalHttpClient.cs [hunk 17]
```diff
@@ -105,12 +108,29 @@ public AdalHttpClient(string uri, CallState callState)
                         serviceEx = new AdalServiceException(tokenResponse.Error, tokenResponse.ErrorDescription,
                             errorCodes, ex);
 
+                        if (ex.WebResponse.StatusCode == HttpStatusCode.BadRequest && tokenResponse.Error == AdalErrorMessage.InteractionRequired)
```
Comment (aadidgit) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: CRITICAL Refactor this code to not nest more than 3 control flow statements. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS134)

**Comment by aadidgit** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by duncanpMS** [comment]
LGTM

File: src/ADAL.PCL/Utilities/InteractionRequiredExceptionDetails.cs [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System.Runtime.Serialization;
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory.Utilities
+{
+    [DataContract]
+    internal class InteractionRequiredExceptionDetails
```
Comment (jennyf19) [inline src/ADAL.PCL/Utilities/InteractionRequiredExceptionDetails.cs]: done.

File: src/ADAL.PCL/Http/AdalHttpClient.cs (hunk 17 above)
Comment (jennyf19) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: done & updated TokenResponse class to include claims value

File: src/ADAL.PCL/Http/AdalHttpClient.cs [hunk 19]
```diff
@@ -101,6 +100,13 @@ public AdalHttpClient(string uri, CallState callState)
                     if (ex.WebResponse != null)
                     {
                         TokenResponse tokenResponse = TokenResponse.CreateFromErrorResponse(ex.WebResponse);
+
+                        // Check for claims challenge response
+                        if (!string.IsNullOrEmpty(tokenResponse.Claims))
```
Comment (aadidgit) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: CRITICAL Refactor this code to not nest more than 3 control flow statements. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS134)

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 20]
```diff
@@ -48,7 +48,9 @@ internal class AcquireTokenInteractiveHandler : AcquireTokenHandlerBase
 
         private readonly UserIdentifier userId;
 
-        public AcquireTokenInteractiveHandler(RequestData requestData, Uri redirectUri, IPlatformParameters parameters, UserIdentifier userId, string extraQueryParameters, IWebUI webUI)
+        private readonly string claims;
+
+        public AcquireTokenInteractiveHandler(RequestData requestData, Uri redirectUri, IPlatformParameters parameters, UserIdentifier userId, string extraQueryParameters, IWebUI webUI, string claims = "")
```
Comment (jennyf19) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: Which values are you referring to?

File: src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs [hunk 21]
```diff
@@ -247,5 +268,53 @@ protected override bool BrokerInvocationRequired()
 
             return false;
         }
+
+        // ProcessClaims will process the claims passed in, either in the
+        // extraQueryParamters or as a string through AcquireToken
+        internal static string ProcessClaims(string extraQueryParameters, string claims)
```
Comment (jennyf19) [inline src/ADAL.PCL/Flows/AcquireTokenInteractiveHandler.cs]: this is removed in a later commit

File: src/ADAL.PCL/Exceptions/AdalClaimChallengeException.cs [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory.Exceptions
+{
+    /// <summary>
+    /// The exception type thrown when a claims challenge error occurs during token acquisition.
+    /// </summary>
+    public class AdalClaimChallengeException : AdalServiceException
```
Comment (jennyf19) [inline src/ADAL.PCL/Exceptions/AdalClaimChallengeException.cs]: This is by design from the design doc...open to discussion for changing the implementation for .NET

File: src/ADAL.PCL/Http/AdalHttpClient.cs [hunk 23]
```diff
@@ -105,12 +104,18 @@ public AdalHttpClient(string uri, CallState callState)
                         serviceEx = new AdalServiceException(tokenResponse.Error, tokenResponse.ErrorDescription,
                             errorCodes, ex);
 
+                        if(ex.WebResponse.StatusCode == HttpStatusCode.BadRequest && tokenResponse.Error == AdalErrorMessage.InteractionRequired)
```
Comment (aadidgit) [inline src/ADAL.PCL/Http/AdalHttpClient.cs]: CRITICAL Refactor this code to not nest more than 3 control flow statements. [rule](http://iddp-sonarqube:9000/coding_rules#rule_key=csharpsquid%3AS134)

**Pull request merged**